File: express-elasticemail-examples/javascript/examples/sub-accounts.js
import "dotenv/config";
import { Configuration, SubAccountsApi } from "@elasticemail/elasticemail-client-ts-axios";

const config = new Configuration({ apiKey: process.env.ELASTICEMAIL_API_KEY });
const subAccountsApi = new SubAccountsApi(config);

// Sub-accounts let you isolate customers or projects with their own API keys and credits.
// Creating one affects billing, so this script only reads unless CREATE_SUBACCOUNT=true.
const createEnabled = process.env.CREATE_SUBACCOUNT === "true";
const subEmail = process.env.SUBACCOUNT_EMAIL || `sub-${Date.now()}@example.com`;

const fail = (step, err) => {
  console.error(`Error (${step}):`, err.response?.status, err.response?.data ?? err.message);
  process.exit(1);
};

try {
  const { data } = await subAccountsApi.subaccountsGet(20, 0);
  console.log(`Sub-accounts (${data.length}):`);
  for (const s of data) {
    console.log(` - ${s.Email} status=${s.Status} credits=${s.EmailCredits} sent=${s.TotalEmailsSent}`);
  }
} catch (err) {
  fail("list sub-accounts", err);
}

if (!createEnabled) {
  console.log("\nSet CREATE_SUBACCOUNT=true to create a sub-account and assign credits.");
  process.exit(0);
}

try {
  const { data } = await subAccountsApi.subaccountsPost({
    Email: subEmail,
    Password: `Tmp-${Math.random().toString(36).slice(2)}-Aa1!`,
    SendActivation: false,
  });
  console.log("\nSub-account created:", data.Email);

  await subAccountsApi.subaccountsByEmailCreditsPatch(subEmail, { Credits: 1000, Notes: "Initial allocation" });
  console.log("Assigned 1000 credits to", subEmail);

  const { data: key } = await subAccountsApi.subaccountsByEmailApikeyGet(subEmail);
  console.log("Sub-account API key retrieved (length):", key.length);
} catch (err) {
  fail("create sub-account", err);
}

File: rust-elasticemail-examples/examples/sub_accounts.rs
use rust_elasticemail_examples::{config, env_or, print_api_error};
use ElasticEmail::apis::sub_accounts_api;
use ElasticEmail::models::{SubaccountEmailCreditsPayload, SubaccountPayload};

#[tokio::main]
async fn main() {
    let config = config();

    // Sub-accounts let you isolate customers or projects with their own API keys and credits.
    // Creating one affects billing, so this script only reads unless CREATE_SUBACCOUNT=true.
    let create_enabled = env_or("CREATE_SUBACCOUNT", "false") == "true";
    let sub_email = env_or("SUBACCOUNT_EMAIL", &format!("sub-{}@example.com", chrono::Utc::now().timestamp_millis()));

    match sub_accounts_api::subaccounts_get(&config, Some(20), Some(0)).await {
        Ok(list) => {
            println!("Sub-accounts ({}):", list.len());
            for s in list {
                println!(
                    " - {} status={:?} credits={} sent={}",
                    s.email.unwrap_or_default(),
                    s.status,
                    s.email_credits.unwrap_or(0),
                    s.total_emails_sent.unwrap_or(0)
                );
            }
        }
        Err(e) => {
            print_api_error("list sub-accounts", &e);
            std::process::exit(1);
        }
    }

    if !create_enabled {
        println!("\nSet CREATE_SUBACCOUNT=true to create a sub-account and assign credits.");
        return;
    }

    let payload = SubaccountPayload {
        send_activation: Some(false),
        ..SubaccountPayload::new(sub_email.clone(), format!("Tmp-{}-Aa1!", uuid::Uuid::new_v4().simple()))
    };
    match sub_accounts_api::subaccounts_post(&config, payload).await {
        Ok(s) => println!("\nSub-account created: {}", s.email.unwrap_or_default()),
        Err(e) => {
            print_api_error("create sub-account", &e);
            std::process::exit(1);
        }
    }

    let credits = SubaccountEmailCreditsPayload {
        notes: Some("Initial allocation".to_string()),
        ..SubaccountEmailCreditsPayload::new(1000)
    };
    match sub_accounts_api::subaccounts_by_email_credits_patch(&config, &sub_email, credits).await {
        Ok(_) => println!("Assigned 1000 credits to {}", sub_email),
        Err(e) => {
            print_api_error("assign credits", &e);
            std::process::exit(1);
        }
    }

    match sub_accounts_api::subaccounts_by_email_apikey_get(&config, &sub_email).await {
        Ok(key) => println!("Sub-account API key retrieved (length): {}", key.len()),
        Err(e) => {
            print_api_error("get sub-account api key", &e);
            std::process::exit(1);
        }
    }
}
